### DISCUSSION TITLE
Queue button shortcut in the Jump To Song window

### DISCUSSION THREAD

**Discussion by lunesar** [audacious-media-player/audacious#81] [Ideas]
Right now when you use "J" to open the "Jump to Song" window you can immediately type in a song name and then either hit enter or click the "Jump" button to close the window and start playing the song. There is also a queue button which is extremely useful if you just want to add the song to the queue and keep the "Jump to Song" window open.

What I would like to add would be that instead of having to use queue button either with the mouse of hitting "Tab" several times, I would like be able to just use "Shift+Enter" to have the same effect as the queue button but still have the Filter dialog still active so I can type in the name of the next song that I would like to queue or jump to.

Here's my proposed change to src/libaudgui/jump-to-track.cc:
from
```
static gboolean keypress_cb (GtkWidget * widget, GdkEventKey * event)
{
    if (event->keyval == GDK_KEY_Escape)
    {
        audgui_jump_to_track_hide ();
        return true;
    }

    return false;
}
```
to
```
static gboolean keypress_cb (GtkWidget * widget, GdkEventKey * event)
{
    if (event->keyval == GDK_KEY_Escape)
    {
        audgui_jump_to_track_hide ();
        return true;
    }
    if ((event->state == GDK_SHIFT_MASK) && (event->keyval == GDK_KEY_Return))
    {
        // Let Shift+Enter act as a short-cut for the Queue button
        do_queue (nullptr);
        return true;
    }

    return false;
}
```
I confirmed that this will compile with `make`. I had to use `autoheader` and `autoconf` before being able to run `./configure` but I eventually got it to where I could `run make` and `make install`. However, I couldn't get my locally compiled instance to open, so I must have something wrong in trying to locally compile it. I was hoping to test it locally before bringing this up, but I couldn't get it to open. Hopefully this code change looks reasonable enough. It would be a nice feature that I would use a lot. Thanks.

**Comment by lunesar** [comment]
Well, suppose that using "Alt+Q" would be another way to hit the Queue button. That's better than tab or the mouse, but I still think "Shift+Enter" would be a better shortcut than "Alt+Q." Currently both the Close and Jump Buttons have shortcuts, "Esc" and "Return," in addition to being able to use "Alt+c" and "Alt+j". I think the Queue button deserves one as well, and I reccomend "Shift+Enter".

**Comment by radioactiveman** [comment]
Thanks @lunesar for the suggestion.

Your code compiled fine but did not work though. :)
You need to use a bitwise AND do detect the Shift key. And for visual feedback it's better to use `gtk_widget_activate()`.

Implemented as f1d2c668ae24cec48be8c057a1d3e2277ad42276 for GTK and 9b78e85d6ab6b977a434ae05350609a710c61a97 for Qt.

**Comment by lunesar** [comment]
Thanks for fixing my bug in the suggestion and making those commits to implement my suggestion, I appreciate it.

I wonder why `event->state == GDK_SHIFT_MASK` didn't work. Is it because event-state also has bits for num_lock, caps_lock, etc? I just wonder if  `event->state & GDK_SHIFT_MASK` would also make it so not just "shift+enter" works, but also "shift+ctl+enter" as well. If so, that's isn't that big of a deal. Just curious on how GTK works for that.

**Comment by radioactiveman** [comment]
Both assumptions are correct. The event state has several bits for different modifier keys. And yes, "Ctrl+Shift+Enter" also triggers the Queue button now.

https://docs.gtk.org/gdk3/flags.ModifierType.html

**Reply by lunesar** [reply]
Ok, good to know. Thanks for the information and the link to the documentation.
